[Keyboard] add aki27/cocot46plus

markstos opened the pull request:
Adds the cocot46plus keyboard by aki27

This is the RP2040 variant.

I've tried to convert a maximal amount from the old config.h and rules.mk to keyboard.json.

I don't regularly work with QMK source code so there is likely room for improvement here. 

## Types of Changes

- [x] Keyboard (addition or update)

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

markstos committed:
[Keyboard] add aki27/cocot46plus

waffle87 commented on keyboards/aki27/cocot46plus/rules.mk:
```diff
@@ -0,0 +1,6 @@
+
+
+# Not sure how to translate these into keyboard.json
+SRC += matrix.c
+POINTING_DEVICE_DRIVER = adns5050
+OLED_DRIVER     = ssd1306
```
```suggestion
SRC += matrix.c
POINTING_DEVICE_DRIVER = adns5050
```

waffle87 commented on keyboards/aki27/cocot46plus/keyboard.json:
```diff
@@ -0,0 +1,152 @@
+    "extrakey": true,
+    "mousekey": true,
+    "nkro": false,
+    "oled": true,
```
If OLEDs are enabled at the keyboard level, there also needs to be some configuration for them.

markstos replied:
I'll remove it from keyboard.json then.

zvecr commented on keyboards/aki27/cocot46plus/cocot46plus.h:
```diff
@@ -0,0 +1,65 @@
+
+extern cocot_config_t cocot_config;
+
+enum cocot_keycodes {
+
+    COCOT_SAFE_RANGE = SAFE_RANGE,
+    CPI_SW,
+    SCRL_SW,
+    ROT_R15,
+    ROT_L15,
+    SCRL_MO,
+    SCRL_TO,
+    SCRL_IN,
+
+};
+
+#define CPI_SW QK_USER_0
+#define SCRL_SW QK_USER_1
+#define ROT_R15 QK_USER_2
+#define ROT_L15 QK_USER_3
+#define SCRL_MO QK_USER_4
+#define SCRL_TO QK_USER_5
+#define SCRL_IN QK_USER_6
```
You should not use user keycodes at the keyboard level

markstos replied:
Thanks, I'll move these to the keymap level.

drashna replied:
`QK_KB_0` and such should be used here rather than `QK_USER_0` and such.

zvecr commented on keyboards/aki27/cocot46plus/cocot46plus.h:
```diff
@@ -0,0 +1,65 @@
+#define SCRL_TO QK_USER_5
+#define SCRL_IN QK_USER_6
+
+bool encoder_update_user(uint8_t index, bool clockwise);
+bool encoder_update_kb(uint8_t index, bool clockwise);
```
```suggestion
```

zvecr commented on keyboards/aki27/cocot46plus/keyboard.json:
```diff
@@ -0,0 +1,136 @@
+    "pointing_device": true,
+    "rgblight": true
+  },
+  "debounce": 5,
```
```suggestion
```

zvecr commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+#include "matrix.h"
+#include "quantum.h"
+
+#if (MATRIX_COLS <= 8)
+#    define print_matrix_header()  print("\nr/c 01234567\n")
+#    define print_matrix_row(row)  print_bin_reverse8(matrix_get_row(row))
+#    define matrix_bitpop(i)       bitpop(matrix[i])
+#    define ROW_SHIFTER ((uint8_t)1)
+#elif (MATRIX_COLS <= 16)
+#    define print_matrix_header()  print("\nr/c 0123456789ABCDEF\n")
+#    define print_matrix_row(row)  print_bin_reverse16(matrix_get_row(row))
+#    define matrix_bitpop(i)       bitpop16(matrix[i])
+#    define ROW_SHIFTER ((uint16_t)1)
+#elif (MATRIX_COLS <= 32)
+#    define print_matrix_header()  print("\nr/c 0123456789ABCDEF0123456789ABCDEF\n")
+#    define print_matrix_row(row)  print_bin_reverse32(matrix_get_row(row))
+#    define matrix_bitpop(i)       bitpop32(matrix[i])
+#    define ROW_SHIFTER  ((uint32_t)1)
+#endif
```
```suggestion
#define ROW_SHIFTER ((uint8_t)1)
```

zvecr commented on keyboards/aki27/cocot46plus/layout-reference.md:
This file should be removed, as it doesnt really add any value.

markstos replied:
I found this very valuable to understand how to create the correct matrix value in keyboard.json. 

Is the thought that now that the matrix `keyboard.json` has been correctly set up, the visual representation of how the keys relate to pins is no longer useful?

zvecr replied:
This is not resolved.

> Is the thought that now that the matrix keyboard.json has been correctly set up, the visual representation of how the keys relate to pins is no longer useful?

Correct. Maintainers have spent a long time removing legacy layout macros from the repo, we do not need them reintroduced in a different form. The file needs to be removed.

markstos replied:
Thanks. I'll remove it.

markstos commented on keyboards/aki27/cocot46plus/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,163 @@
+    .pressed = false
+};
+
+keyevent_t encoder1_cw = {
+    .key = (keypos_t){.row = 4, .col = 5},
+    .pressed = false
+};
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (index == 0) { /* First encoder */
+        if (clockwise) {
+            encoder1_cw.pressed = true;
+            encoder1_cw.time = (timer_read() | 1);
+            action_exec(encoder1_cw);
+        } else {
+            encoder1_ccw.pressed = true;
+            encoder1_ccw.time = (timer_read() | 1);
+            action_exec(encoder1_ccw);
+        }
+    }
+
+    return true;
+}
+
+
+void matrix_scan_user(void) {
+
+    if (encoder1_ccw.pressed) {
+        encoder1_ccw.pressed = false;
+        encoder1_ccw.time = (timer_read() | 1);
+        action_exec(encoder1_ccw);
+    }
+
+    if (encoder1_cw.pressed) {
+        encoder1_cw.pressed = false;
+        encoder1_cw.time = (timer_read() | 1);
+        action_exec(encoder1_cw);
+    }
+
+}
```
I tested removing the code and using the encoder map data structure and that worked fine, so I'm removing this code.

markstos commented:
Thanks for the prompt and helpful feedback. I've applied a number of suggestions. I currently need to pivot to $job but have started getting some build failures for the firmware.

I can investigate more later, but suggestions welcome!

```
 | /usr/bin/avr-ld: /tmp/ccAEZNvn.ltrans0.ltrans.o: in function `.Loc.4948':
 | /home/mark/git/qmk_firmware/quantum/keyboard.c:420:(.text.startup.main+0xe4): undefined reference to `matrix_init'
 | /usr/bin/avr-ld: /tmp/ccAEZNvn.ltrans0.ltrans.o: in function `.Loc.5015':
 | /home/mark/git/qmk_firmware/quantum/bootmagic/bootmagic.c:65:(.text.startup.main+0x1fa): undefined reference to `matrix_scan'
 | /usr/bin/avr-ld: /home/mark/git/qmk_firmware/quantum/bootmagic/bootmagic.c:67:(.text.startup.main+0x210): undefined reference to `matrix_scan'
 | /usr/bin/avr-ld: /tmp/ccAEZNvn.ltrans0.ltrans.o: in function `.Loc.5033':
 | /home/mark/git/qmk_firmware/quantum/bootmagic/bootmagic.c:56:(.text.startup.main+0x216): undefined reference to `matrix_get_row'
 | /usr/bin/avr-ld: /tmp/ccAEZNvn.ltrans0.ltrans.o: in function `.Loc.5195':
 | /home/mark/git/qmk_firmware/quantum/keyboard.c:538:(.text.startup.main+0x49e): undefined reference to `matrix_scan'
 | /usr/bin/avr-ld: /tmp/ccAEZNvn.ltrans0.ltrans.o:/home/mark/git/qmk_firmware/quantum/keyboard.c:541:(.text.startup.main+0x4b6): undefined reference to `matrix_get_row'
 | /usr/bin/avr-ld: /tmp/ccAEZNvn.ltrans0.ltrans.o:/home/mark/git/qmk_firmware/quantum/keyboard.c:553:(.text.startup.main+0x4ce): undefined reference to `matrix_print'
 | /usr/bin/avr-ld: /tmp/ccAEZNvn.ltrans0.ltrans.o:/home/mark/git/qmk_firmware/quantum/keyboard.c:559:(.text.startup.main+0x4d6): undefined reference to `matrix_get_row'
 | /usr/bin/avr-ld: /tmp/ccAEZNvn.ltrans0.ltrans.o: in function `.Loc.5288':
 | /home/mark/git/qmk_firmware/platforms/suspend.c:45:(.text.startup.main+0x59e): undefined reference to `matrix_scan'
 | /usr/bin/avr-ld: /home/mark/git/qmk_firmware/platforms/suspend.c:48:(.text.startup.main+0x5a6): undefined reference to `matrix_get_row'
 | collect2: error: ld returned 1 exit status
```

markstos commented on keyboards/aki27/cocot46plus/keyboard.json:
```diff
@@ -0,0 +1,133 @@
+    "pointing_device": true,
+    "rgblight": true
+  },
+  "matrix_pins": {
+    "cols":  ["F4", "F5", "F6", "F7", "B1", "B3"],
+    "rows": ["D4", "C6", "D7", "E6", "NO_PIN"],
+    "custom": true,
+    "direct": [
+    [ "A00",   "A01",   "A02",   "A03",   "A04",   "A05" ], 
+    [ "A10",   "A11",   "A12",   "A13",   "A14",   "A15" ], 
+    [ "A20",   "A21",   "A22",   "A23",   "A24",   "A25" ], 
+    [ "A30",   "A31",   "A32",   "A33",   "A34",   "A35" ], 
+    [ null,    null,    "A42",   null,   null,    "A45" ], 
+    [ "B00",   "B01",   "B02",   "B03",   "B04",   "B05" ], 
+    [ "B10",   "B11",   "B12",   "B13",   "B14",   "B15" ], 
+    [ "B20",   "B21",   "B22",   "B23",   "B24",   "B25" ], 
+    [ "B30",   "B31",   "B32",   "B33",   "B34",   "B35" ], 
+    [ null,    null,    "B42", null,    null,     "B45" ]  
+    ]
+  },
```
I've tested the keyboard with built with this firmware and it works. What's not correct here?

fauxpark replied:
row-col matrix and direct pins cannot be used simultaneously. The values of this direct pins array do not make any sense anyway. Additionally, since you have a custom `matrix.c` the col and row pin arrays should not be specified in keyboard.json.

markstos replied:
Thanks. Perhaps the keyboard.json validator could throw an error of `cols`/`rows` and `direct` are both  used. Sounds like I can remove them both.

fauxpark commented on keyboards/aki27/cocot46plus/keymaps/default/rules.mk:
This file can be deleted.

fauxpark commented on keyboards/aki27/cocot46plus/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,125 @@
+
+// Defines names for use in layer keycodes and the keymap
+enum layer_number {
+    _BASE = 0,
+    _LOWER = 1,
+    _RAISE = 2,
+    _TRACKBALL = 3,
```
```suggestion
    _BASE,
    _LOWER,
    _RAISE,
    _TRACKBALL
```

fauxpark commented on keyboards/aki27/cocot46plus/keyboard.json:
```diff
@@ -0,0 +1,133 @@
+  "keyboard_name": "cocot46plus",
+  "manufacturer": "aki27",
+  "maintainer": "markstos",
+  "tags": ["unibody","ergo","corne","trackball","encoder","rgb"]
```
```suggestion
  "tags": ["unibody","ergo","corne","trackball","encoder","rgb"],
```

fauxpark replied:
Still outstanding. You can also just run `qmk format-json -i` on this file to have the CLI automatically clean it up.

fauxpark commented on keyboards/aki27/cocot46plus/cocot46plus.h:
```diff
@@ -0,0 +1,63 @@
+
+enum cocot_keycodes {
+
+    COCOT_SAFE_RANGE = SAFE_RANGE,
```
```suggestion
    COCOT_SAFE_RANGE = QK_KB,
```

fauxpark commented on keyboards/aki27/cocot46plus/cocot46plus.c:
```diff
@@ -0,0 +1,295 @@
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "quantum.h"
+#include <math.h>
+#include "cocot46plus.h"
+#include "wait.h"
+#include "debug.h"
+#include <stdio.h>
```
```suggestion
#include "cocot46plus.h"
#include <math.h>
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "matrix.h"
+#include "quantum.h"
```
```suggestion
```

fauxpark replied:
This is still outstanding. Do not include `quantum.h` here unless there is something that directly requires it.

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+
+static void select_row(uint8_t row)
+{
+    setPinOutput(row_pins[row]);
+    writePinLow(row_pins[row]);
```
```suggestion
    gpio_set_pin_output(row_pins[row]);
    gpio_write_pin_low(row_pins[row]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+
+static void unselect_row(uint8_t row)
+{
+    setPinInputHigh(row_pins[row]);
```
```suggestion
    gpio_set_pin_input_high(row_pins[row]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+static void unselect_rows(void)
+{
+    for(uint8_t x = 0; x < MATRIX_ROWS; x++) {
+        setPinInputHigh(row_pins[x]);
```
```suggestion
        gpio_set_pin_input_high(row_pins[x]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+
+static void select_col(uint8_t col)
+{
+    setPinOutput(col_pins[col]);
+    writePinLow(col_pins[col]);
```
```suggestion
    gpio_set_pin_output(col_pins[col]);
    gpio_write_pin_low(col_pins[col]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+
+static void unselect_col(uint8_t col)
+{
+    setPinInputHigh(col_pins[col]);
```
```suggestion
    gpio_set_pin_input_high(col_pins[col]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+static void unselect_cols(void)
+{
+    for(uint8_t x = 0; x < MATRIX_COLS; x++) {
+        setPinInputHigh(col_pins[x]);
```
```suggestion
        gpio_set_pin_input_high(col_pins[x]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+  unselect_rows();
+  unselect_cols();
+  for (uint8_t x = 0; x < MATRIX_COLS; x++) {
+    setPinInputHigh(col_pins[x]);
```
```suggestion
    gpio_set_pin_input_high(col_pins[x]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+    setPinInputHigh(col_pins[x]);
+  }
+  for (uint8_t x = 0; x < MATRIX_ROWS; x++) {
+    setPinInputHigh(row_pins[x]);
```
```suggestion
    gpio_set_pin_input_high(row_pins[x]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+    for(uint8_t col_index = 0; col_index < MATRIX_COLS; col_index++) {
+
+        // Select the col pin to read (active low)
+        uint8_t pin_state = readPin(col_pins[col_index]);
```
```suggestion
        uint8_t pin_state = gpio_read_pin(col_pins[col_index]);
```

fauxpark commented on keyboards/aki27/cocot46plus/matrix.c:
```diff
@@ -0,0 +1,175 @@
+        matrix_row_t last_row_value = current_matrix[tmp];
+
+        // Check row pin state
+        if (readPin(row_pins[row_index]) == 0)
```
```suggestion
        if (gpio_read_pin(row_pins[row_index]) == 0)
```

fauxpark commented on lib/chibios:
Please undo these submodule changes.

markstos commented:
I believe I've addressed the outstanding feedback and did a test flash of the firmware to the keyboard, but afterwards only the left halve and the pointing device register any key events. I'll work on reverting different parts of the changes until I get back to a working state. I expect it was changes related to the matrix.

zvecr commented on keyboards/aki27/cocot46plus/cocot46plus.c:
```diff
@@ -0,0 +1,286 @@
+*/
+
+#include QMK_KEYBOARD_H
+#include "cocot46plus.h"
```
```suggestion
```

markstos replied:
Other keyboards explicitly include their header file. Why is this not needed here?

zvecr replied:
`QMK_KEYBOARD_H` will pull in that header, so its completely pointless.

fauxpark replied:
the `QMK_KEYBOARD_H` include should be removed instead of this one, as the header file name is known.

zvecr replied:
> the QMK_KEYBOARD_H include should be removed instead of this one, as the header file name is known.

This is incorrect, its required for the generated keycodes.

zvecr commented on keyboards/aki27/cocot46plus/rules.mk:
```diff
@@ -0,0 +1,2 @@
+SRC += matrix.c
```
```suggestion
CUSTOM_MATRIX = lite
SRC += matrix.c
```

markstos replied:
Thanks, adding.

zvecr commented on keyboards/aki27/cocot46plus/matrix.c:
This file wont be compiled due to the lack of `CUSTOM_MATRIX = lite` config, and so it will be picking up the one in quantum.

zvecr commented on keyboards/aki27/cocot46plus/cocot46plus.c:
```diff
@@ -0,0 +1,286 @@
+
+    switch (keycode) {
+#ifndef MOUSEKEY_ENABLE
+        // process KC_MS_BTN1~8 by myself
```
> process KC_MS_BTN1~8 by myself

Whats the added value here? It seems to just be duplicating core.

markstos replied:
I'm not sure why that was there. Seems to work OK without it. Removing.

markstos commented:
Thanks for the help again. I've addressed the most recent suggestions and now the right half of the keyboard is working again-- 🚀-- but I noticed the OLED was no longer working. So I added back oled:true, but now the screen displays a random pattern of dots with "506" in the upper left corner. I'll go back through and revert the OLED-related refinements until it works again.

markstos commented:
I understand now what went wrong in the OLED refactoring. There were two issues with the attempt to get rid of snprint. The first was the number of characters printed didn't precisely match, the second was that the old code used an signed int to display an angle of "-30", but the new function, `get_u8_str()` expects an unsigned int.

This keyboard is challenged to have all of its features enabled on a Pro Micro on modern QMK. I've ordered the parts to switch mine over to an RP2040 with more storage. Then we could use snprint without complications. 

I'll convert this a draft PR and re-open once I have something confirmed to work with the RP2040.

markstos commented:
I now have this keyboard working with RP2040 and am targeting that instead of Pro Micro. I was able to get an get an alternative to `sprintf` working, so I reverted to using that, which there is plenty of room for on the RP2040.

zvecr commented on keyboards/aki27/cocot46plus/rules.mk:
```diff
@@ -0,0 +1,7 @@
+SPACE_CADET_ENABLE = no
+# A regular Pro Micro may work if this line is removed
+# but some features may need to be disabled for the firmware to fit in the storage.
+CONVERT_TO=promicro_rp2040
```
Converters should not be enabled at the keyboard level.

(Any existing use in the repo has either already been removed, or is in the process of being removed.)

markstos replied:
Thanks for the feedback.

I tried removing this line and changing the `development_board` in `keyboard.json` from `promicro` to `promicro_rp2040`, but then I get compile errors. What's the proper way to get rid of `CONVERT_TO`?

```
Compiling: keyboards/aki27/cocot46plus/matrix.c                                                    In file included from <command-line>:
./keyboards/aki27/cocot46plus/config.h:30:27: error: 'D4' undeclared here (not in a function)
   30 | #define MATRIX_ROW_PINS { D4, C6, D7, E6, NO_PIN }
      |                           ^~
keyboards/aki27/cocot46plus/matrix.c:20:44: note: in expansion of macro 'MATRIX_ROW_PINS'
   20 | static const pin_t row_pins[MATRIX_ROWS] = MATRIX_ROW_PINS;
      |                                            ^~~~~~~~~~~~~~~
```

fauxpark replied:
You need to do all the other things that the converter does. In this case you will need to redo the pin assignments to match; RP2040 uses `GPx` notation.

markstos commented:
I found the documentation of the GPX notation: https://docs.qmk.fm/platformdev_rp2040#pin-nomenclature

And I found the source code of the converter, but it doesn't show what the mappings to GPx notation are, but some other notation that ends in "U". (Below). 

I found the pin-out for the Pro Micro: https://learn.sparkfun.com/tutorials/pro-micro--fio-v3-hookup-guide/hardware-overview-pro-micro

But this doesn't match either the left or right-hand side of what's in the converter file. 

I'm stuck for now. 

```
// Left side (front)
#define D3 0U
#define D2 1U
//      GND
//      GND
#define D1 2U
#define D0 3U
#define D4 4U
#define C6 5U
#define D7 6U
#define E6 7U
#define B4 8U
#define B5 9U

// Right side (front)
//      RAW
//      GND
//      RESET
//      VCC
#define F4 29U
#define F5 28U
#define F6 27U
#define F7 26U
#define B1 22U
#define B3 20U
#define B2 23U
#define B6 21U

// LEDs (Mapped to QT connector to avoid collisions with button/neopixel)
#define D5 17U
#define B0 16U

```

yudai-nkt commented:
I tried this PR yesterday, and the following commit seems to do the job mentioned in the review: 

yudai-nkt/qmk_firmware@8ec47dbe2e6e4fddaa55068699d4e29efd14fe30

However, this introduce another error:

```sh
$ qmk compile -kb aki27/cocot46plus -km default
Ψ Compiling keymap with gmake -r -R -f builddefs/build_keyboard.mk -s KEYBOARD=aki27/cocot46plus KEYMAP=default KEYBOARD_FILESAFE=aki27_cocot46plus TARGET=aki27_cocot46plus_default INTERMEDIATE_OUTPUT=.build/obj_aki27_cocot46plus_default VERBOSE=false COLOR=true SILENT=false QMK_BIN="qmk"

[snip]

platforms/chibios/drivers/i2c_master.c: In function 'i2c_init':
././platforms/chibios/boards/QMK_PM2040/configs/config.h:17:26: error: 'D0' undeclared (first use in this function)
   17 | #    define I2C1_SCL_PIN D0
      |                          ^~

[snip]

 [ERRORS]
 |
 |
 |
gmake: *** [builddefs/common_rules.mk:373: .build/obj_aki27_cocot46plus_default/i2c_master.o] Error 1
```

The following patch fixes the above error and builds a functioning firmware, but I doubt this is an appropriate fix as it involves modification outside the `keyboard` directory.

```diff
diff --git a/platforms/chibios/boards/QMK_PM2040/configs/config.h b/platforms/chibios/boards/QMK_PM2040/configs/config.h
index f8b46b7fe4..87554f6b83 100644
--- a/platforms/chibios/boards/QMK_PM2040/configs/config.h
+++ b/platforms/chibios/boards/QMK_PM2040/configs/config.h
@@ -11,10 +11,10 @@
 #    define I2C_DRIVER I2CD1
 #endif
 #ifndef I2C1_SDA_PIN
-#    define I2C1_SDA_PIN D1
+#    define I2C1_SDA_PIN GP2
 #endif
 #ifndef I2C1_SCL_PIN
-#    define I2C1_SCL_PIN D0
+#    define I2C1_SCL_PIN GP3
 #endif
```

I asked a similar question at Discord but got no response so far.

fauxpark commented:
The U suffixed values are an intermediate step. Just look at the pinout on the hardware you have...

zvecr commented on keyboards/aki27/cocot46plus/config.h:
```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+// These do not seem to have keyboard.json support yet.
+#define ADNS5050_SCLK_PIN         23U
+#define ADNS5050_SDIO_PIN         8U
+#define ADNS5050_CS_PIN           9U
```
```suggestion
#define ADNS5050_SCLK_PIN         GP23
#define ADNS5050_SDIO_PIN         GP8
#define ADNS5050_CS_PIN           GP9
```

zvecr commented on keyboards/aki27/cocot46plus/rules.mk:
```diff
@@ -0,0 +1,4 @@
+SRC += matrix.c
+CUSTOM_MATRIX = lite
+POINTING_DEVICE_DRIVER = adns5050
+SPACE_CADET_ENABLE = no
```
If required, this should be added to `keyboard.json` features block as `"space_cadet": false`
```suggestion
```

markstos replied:
Now that we are using RP2040, I'll just leave this enabled. Thanks for pointing out this detail.

markstos commented:
The encoder is working now, but the lights are now not.

I suspect the pin number for the ws2812 driver is not right.

markstos commented:
@yudai-nkt Are the lights working you with the latest changes on this branch? Could you check the pins for those?

yudai-nkt commented:
@markstos I've just sent a PR against your fork that fixes my silly mistake. Please take a look when you have time!

markstos commented:
All feedback is now been addressed and all features of the firmware have been tested to work as intended. 

Thanks to everyone for the collaboration.

drashna commented on keyboards/aki27/cocot46plus/keyboard.json:
```diff
@@ -0,0 +1,130 @@
+  "usb": {
+    "device_version": "1.0.0",
+    "pid": "0x0003",
+    "vid": "0x1727"
```
This vendor ID is in use: 
https://www.the-sz.com/products/usbid/index.php?v=0x1727&p=&n=

Generally, the further past 0x4000 you are, the better.

markstos replied:
Thanks for catching that. I've notified the original keyboard designer and will give them a bit to respond, as they have some other keyboards with the same conflicting vendor ID. If I don't hear from them after a while, I guess we should just pick a new ID for this keyboard.

markstos replied:
I did not hear back from the keyboard creator after several attempts on multiple mediums, so I picked a new vendor ID myself and pushed a commit for that.

drashna commented on keyboards/aki27/cocot46plus/cocot46plus.c:
```diff
@@ -0,0 +1,280 @@
+}
+
+
+void matrix_init_kb(void) {
+    // is safe to just read CPI setting since matrix init
+    // comes before pointing device init.
+    cocot_config.raw = eeconfig_read_kb();
+    if (cocot_config.cpi_idx > CPI_OPTION_SIZE) // || cocot_config.scrl_div > SCRL_DIV_SIZE || cocot_config.rotation_angle > ANGLE_SIZE)
+    {
+        eeconfig_init_kb();
+    }
+    matrix_init_user();
+}
```
```suggestion
void keyboard_post_init_kb(void) {
    // is safe to just read CPI setting since matrix init
    // comes before pointing device init.
    cocot_config.raw = eeconfig_read_kb();
    if (cocot_config.cpi_idx > CPI_OPTION_SIZE) // || cocot_config.scrl_div > SCRL_DIV_SIZE || cocot_config.rotation_angle > ANGLE_SIZE)
    {
        eeconfig_init_kb();
    }
    keyboard_post_init_user();
}
```

sigprof replied:
Why this had been marked as resolved without either changing the code or writing some comment?

The suggestion here is not 100% correct, because `keyboard_post_init_kb()` actually gets called after pointing device init. However, doing eeconfig handling from `matrix_init_kb()` is not 100% correct too, because it gets called before `quantum_init()`, which is where eeconfig init may be performed, so anything you do with eeconfig here may get overwritten (although the core will call `eeconfig_init_kb()` in that case, so you just would do some duplicate work).

drashna commented on keyboards/aki27/cocot46plus/cocot46plus.c:
```diff
@@ -0,0 +1,280 @@
+    char cpi[5];
+    char scroll_div[3];
+    char angle[4];
+    snprintf(cpi, 5, "%4d", cpi_array[cocot_config.cpi_idx]);
+    snprintf(scroll_div, 3, "%2d", scrl_div_array[cocot_config.scrl_div]);
+    snprintf(angle, 4, "%3d",  angle_array[cocot_config.rotation_angle]);
```
just a heads up, you can use `get_u8_str(x, char)` and `get_u16_str(x, char)` to generate a 3 and 5 character array with the number values.

markstos replied:
Thanks for the tip! I'm going to pass in this case, but I'll keep it in the mind for the future.

drashna commented on keyboards/aki27/cocot46plus/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,134 @@
+
+
+#ifdef OLED_ENABLE
+bool oled_task_user(void) {
```
This needs to be moved to the kb level, and as:
```c
bool oled_task_kb(void) {
    if (!oled_task_user()) {
        return false;
    }
    // ... rest of code
}
```

markstos replied:
Thanks for the feedback, @drashna 

I've pushed a commit that does this now.

drashna commented on keyboards/aki27/cocot46plus/cocot46plus.c:
```diff
@@ -0,0 +1,289 @@
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include QMK_KEYBOARD_H
```
```suggestion
#include "cocot46plus.h"
```

markstos replied:
After this proposed change was made, the keyboard no long compiles:

```
Compiling: keyboards/aki27/cocot46plus/cocot46plus.c                                               keyboards/aki27/cocot46plus/cocot46plus.c: In function 'process_record_kb':
keyboards/aki27/cocot46plus/cocot46plus.c:135:20: error: 'CPI_SW' undeclared (first use in this function)
  135 |     if (keycode == CPI_SW && record->event.pressed) {
      |                    ^~~~~~
keyboards/aki27/cocot46plus/cocot46plus.c:135:20: note: each undeclared identifier is reported only once for each function it appears in
keyboards/aki27/cocot46plus/cocot46plus.c:141:20: error: 'SCRL_SW' undeclared (first use in this function)
  141 |     if (keycode == SCRL_SW && record->event.pressed) {
      |                    ^~~~~~~
keyboards/aki27/cocot46plus/cocot46plus.c:146:20: error: 'ROT_R15' undeclared (first use in this function)
  146 |     if (keycode == ROT_R15 && record->event.pressed) {
      |                    ^~~~~~~
keyboards/aki27/cocot46plus/cocot46plus.c:151:20: error: 'ROT_L15' undeclared (first use in this function)
  151 |     if (keycode == ROT_L15 && record->event.pressed) {
      |                    ^~~~~~~
keyboards/aki27/cocot46plus/cocot46plus.c:156:20: error: 'SCRL_IN' undeclared (first use in this function)
  156 |     if (keycode == SCRL_IN && record->event.pressed) {
      |                    ^~~~~~~
keyboards/aki27/cocot46plus/cocot46plus.c:161:20: error: 'SCRL_TO' undeclared (first use in this function)
  161 |     if (keycode == SCRL_TO && record->event.pressed) {
      |                    ^~~~~~~
keyboards/aki27/cocot46plus/cocot46plus.c:167:20: error: 'SCRL_MO' undeclared (first use in this function)
  167 |     if (keycode == SCRL_MO) {
```

fauxpark replied:
That's because these keycodes are generated by the CLI from the list in `keyboard.json`. I guess th4 best solution here is to replace the `quantum.h` include in this header with `QMK_KEYBOARD_H`.

markstos replied:
Then there is nothing to do, as the change to replace "QMK_KEYBOARD_H" was not added to this branch. Marking as resolved.

drashna commented on keyboards/aki27/cocot46plus/cocot46plus.c:
```diff
@@ -0,0 +1,289 @@
+    return false;
+}
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
```
```suggestion
oled_rotation_t oled_init_kb(oled_rotation_t rotation) {
```

drashna commented on keyboards/aki27/cocot46plus/matrix.c:
Could you run a format pass on this file?  (`qmk format-c (name)` will accomplish that)

markstos commented:
This appears to be right at the finish line. Is there something I can do to help it along?

sigprof commented on keyboards/aki27/cocot46plus/cocot46plus.h:
```diff
@@ -0,0 +1,41 @@
+
+#include "quantum.h"
+
+typedef union {
+    uint32_t raw;
+    struct {
+        uint8_t cpi_idx;
+        uint8_t scrl_div;
+        uint8_t rotation_angle;
+        int8_t scrl_inv;
+        bool scrl_mode;
+        report_mouse_t last_mouse;
+    };
+} cocot_config_t;
```
The inner struct is obviously larger than `uint32_t`, so only the first 4 fields get saved in the EEPROM, even if the appearance of this union suggests otherwise. To avoid confusion, you should split this into separate data items for the persistent and runtime-only data.

markstos replied:
I'm adding code commits to clarify that `scrl_mode` and `last_mouse` are not saved to EEPROM

sigprof commented on keyboards/aki27/cocot46plus/cocot46plus.c:
```diff
@@ -0,0 +1,289 @@
+    cocot_config.scrl_mode = false;
+    eeconfig_update_kb(cocot_config.raw);
+    eeconfig_init_user();
+    adns5050_write_reg(0x22, 0b10000 | 0x80);
```
What does this do, and is this function safe to call from here? Note that `eeconfig_init_kb()` can be called from `eeconfig_init_quantum()` rather early during startup.

markstos replied:
> What does this do?

A number of keyboards define their own `eeconfig_init_kb()` function.
https://docs.qmk.fm/feature_eeprom#eeconfig-function-documentation

Search in this repo to see examples. It's normal that they call `eeconfig_update_kb()` and `eeconfig_init_user()`. 
As a random selection, ` keyboards/keebio/quefrency/rev1/rev1.c` is one file that does this. 

> and is this function safe to call from here?

Yes. It's working fine on this keyboard and there are 42  other QMK keyboards that define their own `eeconfig_init_kb()` function.  

Maybe you are asking specifically about `adns5050_write_reg()`. That method is part of the ADNS 5050 Sensor driver provided by QMK. It's defined in `drivers/sensors/adns5050.*`: https://docs.qmk.fm/features/pointing_device#adns-5050-sensor

Curiously, This keyboard is the *only* keyboard in the QMK repo that's calling this method. Probably because it's a LED trackball sensor, and more accurate laser sensors are available for just a few dollars more. Now I want to upgrade this part...

zvecr replied:
You understand the initial question is "why are you calling `adns5050_write_reg` from within `eeconfig_init_kb`"?

Explaining what core functions are to maintainers is rather bold given they will already have domain knowledge.

Please focus on providing the justification as to why `adns5050_write_reg` must be called from within eeconfig init, or remove/relocate it to another more general part of the init process.

markstos commented:
Latest feedback has been addressed.

markstos commented:
Latest feedback has been addressed.
